dhis-2/dhis-api/src/main/java/org/hisp/dhis/dataintegrity/FlattenedDataIntegrityReport.java

```diff
@@ -151,151 +155,144 @@
     {
-        HashMap<String, String> newMap = new HashMap<>( map.size() );
-
-        for ( Map.Entry<? extends IdentifiableObject, String> entry : map.entrySet() )
-        {
-            newMap.put( defaultIfNull( entry.getKey() ), entry.getValue() );
-        }
-
-        return newMap;
+        return map == null
+            ? null
+            : map.entrySet().stream().collect( toUnmodifiableMap(
+                e -> defaultIfNull( e.getKey() ), Entry::getValue ) );
     }
 
-    private Map<String, Collection<String>> transformMapOfCollections(
+    private Map<String, Collection<String>> mapToDisplayNameOrId(
         Map<? extends IdentifiableObject, ? extends Collection<? extends IdentifiableObject>> map )
     {
-        HashMap<String, Collection<String>> newMap = new HashMap<>();
-
-        for ( var entry : map.entrySet() )
-        {
-            newMap.put( defaultIfNull( entry.getKey() ), new HashSet<>( transformCollection( entry.getValue() ) ) );
-        }
-
-        return newMap;
+        return map == null
+            ? null
+            : map.entrySet().stream().collect( toUnmodifiableMap(
+                e -> defaultIfNull( e.getKey() ),
+                e -> mapToListOfDisplayNameOrUid( e.getValue() ) ) );
     }
 
-    private List<String> transformCollection( Collection<? extends IdentifiableObject> collection )
+    private List<String> mapToListOfDisplayNameOrUid( Collection<? extends IdentifiableObject> collection )
     {
-        List<String> newCollection = new ArrayList<>( collection.size() );
-
-        for ( IdentifiableObject o : collection )
-        {
-            newCollection.add( StringUtils.defaultIfBlank( o.getDisplayName(), o.getUid() ) );
-        }
-
-        return newCollection;
+        return collection == null
+            ? null
+            : collection.stream()
+                .map( o -> defaultIfBlank( o.getDisplayName(), o.getUid() ) )
+                .collect( toUnmodifiableList() );
     }
 
-    private SortedMap<String, Collection<String>> transformSortedMap(
+    private SortedMap<String, Collection<String>> mapToSortedDisplayNameOrUid(
         SortedMap<? extends IdentifiableObject, ? extends Collection<? extends IdentifiableObject>> map )
     {
-        SortedMap<String, Collection<String>> newMap = new TreeMap<>();
-
-        for ( var entry : map.entrySet() )
-        {
-            newMap.put( defaultIfNull( entry.getKey() ), transformCollection( entry.getValue() ) );
-        }
-
-        return newMap;
+        return map == null
+            ? null
+            : map.entrySet().stream().collect( toMap(
+                e -> defaultIfNull( e.getKey() ),
+                e -> mapToListOfDisplayNameOrUid( e.getValue() ),
+                ( v1, v2 ) -> {
+                    Collection<String> both = new ArrayList<>();
+                    both.addAll( v1 );
+                    both.addAll( v2 );
+                    return both;
```

Comment:
FYI: This might look slightly more complex but it also does handle a case where the old code would just have replace an already mapped key with a new value. This now will aggregate the values which seems more correct in this context.

Reply:
That sounds good. Though I think that code might not be entirely easy to read. Any chance we could make it a bit easier to understand what happens here?

Reply:
Extracted a helper method. Should be more easy to read and clear now.